### PR TITLE
Ignore FID field when merging vector layers to geopackage

### PR DESCRIPTION
**Pull request by m-kuhn**
When merging vector layers which have a field called "fid", the ogr gpkg writer will use this as PK and assume it's unique. Since the fids come from different layers, it's likely not unique.

This commit just drops all the fids and regenerates them.

Questions: is this approach ok?

Preliminary code for early preview.

### CODE REVIEW

**Comment by nyalldawson** [comment]
I forgot about this pr, but please don't merge till I've had a chance to review!

**Comment by nyalldawson** [comment]
@m-kuhn like we discussed, i'm not 100% on this approach. I think i'd prefer us just to always ignore any existing fid fields and always create new ones. Because I also see a similar issue when:

- load a gml file (such as processing test data)
- run ANY alg (try buffer) and save to gpkg
- you get the error "Creation of field fid failed (OGR error: Wrong field type for fid)", because the existing fid field isn't compatible with gpkg's fid requirements

So I'd tend towards just ALWAYS dropping existing fids and letting the output format recreate them in whatever format meets its needs.

But if we do have to go with this current approach, can we use a flags based approach instead of a single-use bool?

**Comment by nirvn** [comment]
Another fid issue I've ran into: multi to single parts will often trigger a "error, fid not unique" when saving to gpkg.

**Comment by m-kuhn** [comment]
@nyalldawson I am not sure I completely understand your proposal. Are you talking about geopackage or about any output format? I assume only geopackage, because another format may suddenly have a hardcoded name "gid" as primary key.

Also for geopackage I'm not sure we should just "always" do it. I think the way we present the attribute as a normal field inside QGIS doesn't warrant for this. If we really want the fid to always be disposable we need to remove it from the attribute table and only use it as `$id` (to be clear: I'm not proposing that). So whenever possible I would like to preserve the attribute fid, just like we preserve any other attribute. I'm not sure where the buffer problem is you mention, but that would be a case where we have one input feature to one output feature (normally) and having the fid preserved to join the two features later on could be handy in some cases. So as long as we don't have a reason to regenerate the fid (like with merge features or multi to single) we should IMO just keep our hands away from them.

Flags: no problem for me!

**Comment by m-kuhn** [comment]
> Here's another argument for always dropping existing fid fields: https://issues.qgis.org/issues/20207 (at least I think that's also the same root cause)

Isn't that an issue with duplicate attribute names in general, not specific to fid?

> Yes, geopackage only, where fid has a very special meaning and is basically an "internal" detail. (In my understanding)

:+1: 

> The issue is because the input fid values aren't numeric, so can't be written to geopackage. I don't think there's any solution here except to drop them (or rename them to fid_1 and let ogr create a new fid field).

Why does buffering cause non-numeric fids?

Anyway ... the proposed solution allows for any algorithm to specify that they want new fids, but does not force them to (which would be your proposal if I understand correctly). If buffer has a problem with fids, it can just specify the flag. Just like all the others where multiple tables are merged (join*) or existing features are duplicated (multipart to singlepart et al). To me this sounds way more flexible.

**Comment by nyalldawson** [comment]
> Why does buffering cause non-numeric fids?

It doesn't. It's just an example -- but ANY algorithm fails when saving to geopackage and the input has non-numeric fids. Load any GML dataset from the processing test data folder and you'll see -- buffer was just an easy example to use.

> Isn't that an issue with duplicate attribute names in general, not specific to fid?

Well, the algorithm already has code in place here to avoid duplicate attribute names. I'd need to dig deeper to work out what's happening here, but my immediate suspicion is that it's again a consequence of not dropping the input fids. That's just a hunch ;)

**Comment by m-kuhn** [comment]
> It doesn't. It's just an example -- but ANY algorithm fails when saving to geopackage and the input has non-numeric fids. Load any GML dataset from the processing test data folder and you'll see -- buffer was just an easy example to use.

Ah well, that's a different issue. We can also force regeneration of fids on gpkg with non-integer fids.

What I'm trying to avoid is this scenario:

* you have a geopackage with numeric fid and you run a buffer on it
* the output fids get regenerated and happen to be identical to the original fids (because in most cases that's what's going to happen)
* you rely on this behavior (because it sounds reasonable) and build a workflow around it that maps output to input features via the field fid
* at some point, some features in the input layer are deleted, the fids no longer match, the assumption goes boom and so does the nicely designed workflow

**Comment by m-kuhn** [comment]
What this pull request tries to solve is not *anything that might be related to ogr fids*, only that algorithms can specify that they are not able to preserve unique values and the sink can take measures to handle this situation gracefully - which happens to affect gpkg fids.

File: src/core/processing/qgsprocessingutils.cpp [hunk 1]
```diff
@@ -443,14 +444,27 @@ QgsFeatureSink *QgsProcessingUtils::createFeatureSink( QString &destination, Qgs
...
-    parseDestinationString( destination, providerKey, uri, layerName, format, options, useWriter );
+    parseDestinationString( destination, providerKey, uri, layerName, format, adaptedCreateOptions, useWriter );
+
+    QgsFields newFields = fields;
+    // If the result is a geopackage layer and there is already a field name FID requested which
+    // might contain duplicates, make sure to generate a new field with a unique name instead
+    // that will be filled by ogr with unique values.
+    if ( options.testFlag( QgsProcessingUtils::FidMayContainDuplicates ) && format == QLatin1String( "GPKG" ) && !adaptedCreateOptions.contains( QStringLiteral( "FID" ) ) )
+    {
+      int fidIdx = fields.lookupField( QStringLiteral( "FID" ) );
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingutils.cpp]: Ok, moving on with the discussion :)

Am I missing something here, or won't the attributes added to this sink contain the existing fid value, and accordingly be one row out of sync with the actual sink fields?

Comment (m-kuhn) [reply src/core/processing/qgsprocessingutils.cpp]: Correct. As soon as it's run with more fields it gets out of sync.

Will have to think of something else... But that's gonna be 3.4.1

**Comment by m-kuhn** [comment]
Adjusted this pull request.
It's now all down to the feature sinks (like QgsVectorFileWriter or QgsVectorLayerExporter) to properly deal with the situation based on a flag `QgsFeatureSink::SinkFlag::RegeneratePrimaryKey` which can be specified by whoever thinks it's a good idea to specify this.

File: src/core/qgsvectorlayerexporter.cpp [hunk 2]
```diff
@@ -120,6 +121,21 @@ QgsVectorLayerExporter::QgsVectorLayerExporter( const QString &uri,
     return;
   }
 
+  // If the result is a geopackage layer and there is already a field name FID requested which
+  // might contain duplicates, make sure to generate a new field with a unique name instead
+  // that will be filled by ogr with unique values.
+
+  // HACK sorry
+  if ( sinkOptions.testFlag( QgsFeatureSink::SinkFlag::RegeneratePrimaryKey ) && uri.endsWith( QLatin1String( ".gpkg" ) ) )
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerexporter.cpp]: I think you'll need to use decodeUri here -- couldn't a layername etc be sitting on the end of the uri?

File: src/core/qgsvectorlayerexporter.h [hunk 3]
```diff
@@ -105,7 +105,8 @@ class CORE_EXPORT QgsVectorLayerExporter : public QgsFeatureSink
                             QgsWkbTypes::Type geometryType,
                             const QgsCoordinateReferenceSystem &crs,
                             bool overwrite = false,
-                            const QMap<QString, QVariant> &options = QMap<QString, QVariant>() );
+                            const QMap<QString, QVariant> &options = QMap<QString, QVariant>(),
+                            QgsFeatureSink::SinkFlags sinkOptions = nullptr );
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerexporter.h]: sinkOptions -> sinkFlags ?

**Comment by nyalldawson** [comment]
I like this approach much better.

It started me wondering if this could/(should also?) be a new flag in QgsFeatureSink::Flag ? I'm guessing it would need to be in addition, because of the setup work required here. But would we also want a per-feature option to regenerate just that features' primary key? Not sure.

File: src/core/qgsvectorlayerexporter.cpp (hunk 2 above)
Comment (m-kuhn) [reply src/core/qgsvectorlayerexporter.cpp]: I can try. Those uris in all the variants are a miracle to me :/

**Comment by m-kuhn** [comment]
Merge?

**Comment by m-kuhn** [comment]
> It started me wondering if this could/(should also?) be a new flag in QgsFeatureSink::Flag ? I'm guessing it would need to be in addition, because of the setup work required here. But would we also want a per-feature option to regenerate just that features' primary key? Not sure.

I was thinking about that too and was hesitating to add this as a new flag. But the way it's currently implemented (modifying a member variable) makes it more stable as a sink-based flag and not per-feature flag. I also think that's what's required mostly (I'm not sure if e.g. the fastinsert flag is somewhere actually used individually per feature). Last but not least, to recalculate a fid for a single feature, you can still just specify an invalid qvariant in this case.

File: src/core/qgsvectorlayerexporter.cpp [hunk 4]
```diff
@@ -120,6 +121,22 @@ QgsVectorLayerExporter::QgsVectorLayerExporter( const QString &uri,
     return;
   }
 
+  // If the result is a geopackage layer and there is already a field name FID requested which
+  // might contain duplicates, make sure to generate a new field with a unique name instead
+  // that will be filled by ogr with unique values.
+
+  // HACK sorry
+  const QString path = QgsProviderRegistry::instance()->decodeUri( QStringLiteral( "ogr" ), uri ).value( QStringLiteral( "path" ) ).toString();
+  if ( sinkFlags.testFlag( QgsFeatureSink::SinkFlag::RegeneratePrimaryKey ) && path.endsWith( QLatin1String( ".gpkg" ) ) )
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerexporter.cpp]: path.endsWith( QLatin1String( ".gpkg" ), Qt::CaseInsensitive )

File: src/core/qgsvectorlayerexporter.cpp [hunk 5]
```diff
@@ -120,6 +121,22 @@ QgsVectorLayerExporter::QgsVectorLayerExporter( const QString &uri,
     return;
   }
 
+  // If the result is a geopackage layer and there is already a field name FID requested which
+  // might contain duplicates, make sure to generate a new field with a unique name instead
+  // that will be filled by ogr with unique values.
+
+  // HACK sorry
+  const QString path = QgsProviderRegistry::instance()->decodeUri( QStringLiteral( "ogr" ), uri ).value( QStringLiteral( "path" ) ).toString();
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerexporter.cpp]: Could we get a specific unit test covering this?

Comment (m-kuhn) [reply src/core/qgsvectorlayerexporter.cpp]: "this"?

Comment (m-kuhn) [reply src/core/qgsvectorlayerexporter.cpp]: done

**Comment by m-kuhn** [comment]
That one is tricky. It's a "feature based algorithm" which normally don't need this (most of them are one in, one out)

**Comment by nirvn** [comment]
@m-kuhn , yeah, except this one is one in, very likely two or more out :) that said, if there's more work to be done to code a feature-based alg solution, ignore that one in this PR :)

The first time I ever encountered this fid issue was with that algorithm.

**Comment by nirvn** [comment]
@m-kuhn , cheers :)

**Comment by m-kuhn** [comment]
I thought feature based algorithms are also the ones which are e.g. predetermined for "update in place" which doesn't make much sense with multiple out. So my last comment was mostly based on my surprise than on actual technical difficulties.

**Comment by rduivenvoorde** [comment]
I think this issue is related isn't it?
https://issues.qgis.org/issues/17899

That is because the original data has a fid of type Real/Double...
I think this PR could potentially close a lot of issues :-)

**Pull request merged**